godot-codegen/src/models/domain/enums.rs

```diff
@@ -0,0 +1,164 @@
+
+use proc_macro2::{Ident, Literal, TokenStream};
+use quote::{quote, ToTokens};
+
+pub struct Enum {
+    pub name: Ident,
+    pub godot_name: String,
+    pub is_bitfield: bool,
+    pub enumerators: Vec<Enumerator>,
+}
+
+impl Enum {
+    /// Which derives should be implemented for this enum.
+    pub fn derives(&self) -> Vec<Ident> {
+        let mut derives = vec!["Copy", "Clone", "Eq", "PartialEq", "Hash", "Debug"];
+
+        if self.is_bitfield {
+            derives.push("Default");
+        }
+
+        derives.into_iter().map(ident).collect()
+    }
+
+    /// The type we use to represent values of this enum.
+    pub fn ord_type(&self) -> Ident {
+        if self.is_bitfield {
+            ident("u64")
+        } else {
+            ident("i32")
+        }
+    }
+
+    /// Returns all unique enumerator ords, sorted.
+    ///
+    /// Returns `None` if `self` is a bitfield.
+    pub fn unique_ords(&self) -> Option<Vec<i32>> {
+        let mut unique_ords = self
+            .enumerators
+            .iter()
+            .map(|enumerator| enumerator.value.as_enum())
+            .collect::<Option<Vec<i32>>>()?;
+
+        unique_ords.sort();
+        unique_ords.dedup();
+
+        Some(unique_ords)
+    }
+
+    /// Returns tokens representing the engine trait this enum should implement.
+    pub fn engine_trait(&self) -> TokenStream {
+        if self.is_bitfield {
+            quote! { crate::obj::EngineBitfield }
+        } else {
+            quote! { crate::obj::EngineEnum }
+        }
+    }
+
+    /// Returns the maximum index of an indexable enum.
+    ///
+    /// Return `None` if `self` isn't an indexable enum. Meaning it is either a bitfield, or it is an enum that can't be used as an index.
+    pub fn count_index_enum(&self) -> Option<usize> {
+        if self.is_bitfield {
+            return None;
+        }
+
+        // Sort by ordinal value. Allocates for every enum in the JSON, but should be OK (most enums are indexable).
+        let enumerators = {
+            let mut enumerators = self.enumerators.clone();
+            enumerators.sort_by_key(|v| v.value.to_i64());
+            enumerators
+        };
+
+        // Highest ordinal must be the "MAX" one.
+        // The presence of "MAX" indicates that Godot devs intended the enum to be used as an index.
+        // The condition is not strictly necessary and could theoretically be relaxed; there would need to be concrete use cases though.
+        let last = enumerators.last()?; // If there isn't a last we can assume it shouldn't be used as an index.
+
+        if !last.godot_name.ends_with("_MAX") {
+            return None;
+        }
+
+        // The rest of the enumerators must be contiguous and without gaps (duplicates are OK).
+        let mut last_value = 0;
+
+        for enumerator in enumerators.iter() {
+            let current_value = enumerator.value.to_i64();
+
+            if !(current_value == last_value || current_value == last_value + 1) {
```

Comment:
Imo easier to understand:
```suggestion
            if e_value != last_value && e_value != last_value + 1 {
```
(it was like this before, but `e_value` on the LHS is even better)

Reply:
i dont think `e_value` is better than `current_value` here, personally i do find the first one "if `current_value` is neither the last value nor the last value + 1" vs "if `current_value` is different from both the last value and last value + 1" but i can change it

Reply:
Sorry for the misunderstanding, my comment was about the order of operands and the boolean simplification. I agree that `current_value` is better, missed that in my suggestion :slightly_smiling_face: 

So my suggestion would be:
```
            if current_value != last_value && current_value != last_value + 1 {
```

Simply because the expression has one less indirection (parentheses + `!`). But I agree it's a bit matter of style...